reciprocalspaceship/dataset.py

```diff
@@ -878,33 +879,75 @@ def compute_multiplicity(self, inplace=False, include_centering=True):
         return self
 
     @inplace
-    def assign_resolution_bins(self, bins=20, inplace=False, return_labels=True):
+    def assign_resolution_bins(
+        self,
+        bins=20,
+        inplace=False,
+        return_labels=True,
+        format_str=".2f",
+        return_edges=False,
+    ):
         """
         Assign reflections in DataSet to resolution bins.
 
+        Notes
+        -----
+        - If bin edges are provided, any reflections outside of the specified range
+          are dropped.
+
         Parameters
         ----------
-        bins : int
-            Number of bins
+        bins : int, list, or np.ndarray
+            Number of bins or bin edges to use when assigning resolution bins. If
+            bin edges are provided, they must be monotonic (default: 20)
         inplace : bool
-            Whether to add the column in place or return a copy
+            Whether to add the column in place or return a copy (default: False)
         return_labels : bool
             Whether to return a list of labels corresponding to the edges
-            of each resolution bin
+            of each resolution bin (default: True)
+        format_str : str
+            Format string for constructing bin labels
+        return_edges : bool
+            Whether to return bin edges that define the resolution bin boundaries.
+            The bin edges are returned as a 1-dimensional array with `bins + 1` entries
+            (default: False)
 
         Returns
         -------
-        (DataSet, list) or DataSet
+        (DataSet, list), (DataSet, ndarray), (DataSet, list, ndarray) or DataSet
         """
         dHKL = self.compute_dHKL()["dHKL"]
 
-        assignments, labels = bin_by_percentile(dHKL, bins=bins, ascending=False)
-        self["bin"] = DataSeries(assignments, dtype="I", index=self.index)
+        if isinstance(bins, int):
+            assignments, edges = bin_by_percentile(dHKL, bins=bins, ascending=False)
+            self.loc[:, "bin"] = DataSeries(assignments, dtype="I", index=self.index)
+        else:
+            mask = (dHKL >= min(bins)) & (dHKL <= max(bins))
+            assignments = assign_with_binedges(dHKL[mask], bin_edges=bins)
+            edges = np.array(bins)
+
+            # Drop reflections outside of resolution range
+            result = self.loc[mask].copy()
```

Comment:
ideally you would not make a copy in the case of `inplace=True`.

Reply:
While I agree in theory, I don't think it's possible to do so here without hitting a `SettingWithCopyWarning`. Let me know if you can think of a way to do so here -- I've tried all permutations I could think of.

Reply:
nvm -- I _think_ I got it

Reply:
resolved with new commits